src/alexandria3k/data_sources/uspto.py

```diff
@@ -18,71 +18,529 @@
 #
 """Patent grant bibliographic (front page) text data (JAN 1976 - present)"""
 
+import zipfile
+import os
+
+from alexandria3k.common import fail, warn
+from alexandria3k.data_source import (
+    CONTAINER_INDEX,
+    DataSource,
+    StreamingCachedContainerTable,
+)
+
+from alexandria3k import perf
+from alexandria3k.file_xml_cache import get_file_cache
 from alexandria3k.db_schema import ColumnMeta, TableMeta
 
 # Bulk data can be found here. https://bulkdata.uspto.gov
 # Patent Grant Bibliographic (Front Page) Text Data (JAN 1976 - PRESENT)
-DEFAULT_SOURCE = None
+DEFAULT_SOURCE = (
+    "path/to/tests/data/April 2023 Patent Grant Bibliographic Data"
+)
 
+# Delimiter for extracting concatenated XML files.
+XML_DELIMITER = '<?xml version="1.0" encoding="UTF-8"?>'
 
 # Dataset Description — Patent grant full-text data (no images)
 # JAN 1976 — present Automated Patent System (APS)
 # Contains the full text of each patent grant issued weekly
 # Tuesdays from January 1, 1976, to present excludes images/drawings and reexaminations.
 # https://developer.uspto.gov/product/patent-grant-bibliographic-dataxml
 
+
+class ErrorElement:
+    """A class used for representing error elements"""
+
+    def find(self, _path):
+        """Placeholder to the actual XML tree element."""
+        return None
+
+    def findall(self, _path):
+        """Placeholder to the actual XML tree element."""
+        return None
+
+
+def get_element(tree, path):
+    """Return the text value of the specified element path of the given
+    tree."""
+    element = tree.find(path)
+    # print(element)
+    if element is None:
+        return None
+    return element.text
+
+
+def get_attribute(tree, attr, path=None):
+    """Return the value of the specified attribute. If no path is
+    given searches on root, if specified searches on the specified
+    element."""
+
+    if path:
+        element = tree.find(path)
+        return element.get(attr) if element is not None else None
+    return tree.get(attr) if tree.get(attr) is not None else None
+
+
+def getter(path):
+    """Return a function to return an element with the specified
+    path from a given tree."""
+    return lambda tree: get_element(tree, path)
+
+
+def agetter(attr_name, path=None):
+    """Return a function to return an attribute with the specified
+    name."""
+    return lambda tree: get_attribute(tree, attr_name, path)
+
```

Comment:
I see some commonality with orcid.py. Should some common functions be moved to a helper xml.py file, and documented as part of the plugin development API?